marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/entities/entity-search-lib.sjs

```diff
@@ -535,11 +540,58 @@ function isHubEntityInstance(docUri) {
   return isHubEntityInstance;
 }
 
+function getRecordHistory(docUri) {
+  const history = [];
+  const provResults = xdmp.invokeFunction(function () {
+    const match = {
+      attributes: {
+        location: docUri
+      }
+    };
+    const out = {
+      dateTime: '?',
+      relations: {'associatedWith': '?', 'attributedTo': '?'}
+    };
+    const kvPattern = ps.opTriplePattern(match, out);
+    return op.fromTriples(kvPattern).result();
+  }, { 'database' : xdmp.database(datahub.config.JOBDATABASE)}).toArray();
+
+  if(provResults.length == 0) {
+    const metadata = xdmp.documentGetMetadata(docUri);
+    if(metadata) {
+      const currentObject = {};
+      currentObject.updatedTime = metadata.datahubCreatedOn ? metadata.datahubCreatedOn : undefined;
+      currentObject.flow = metadata.datahubCreatedInFlow ? metadata.datahubCreatedInFlow : undefined;
+      currentObject.step = metadata.datahubCreatedByStep ? metadata.datahubCreatedByStep : undefined;
+      currentObject.user = metadata.datahubCreatedBy ? metadata.datahubCreatedBy : undefined;
+      history.push(currentObject);
+    }
+  } else {
+    const flowNames = fn.collection("http://marklogic.com/data-hub/flow").toArray().map(flow => JSON.stringify(flow.toObject().name));
```

Comment:
Getting all objects in a collection is generally not a safe practice. While we don't expect there to be hundreds of flows or steps, we can't prevent that. 

Instead, try the following approach:

1. Iterate through provResults first, building up the history array and also building arrays of flow and step names. 
1. Take all the flow names and use those to query for only the flow objects you need. 
1. Take all the step names and use those to query for only the steps / step definitions that you need. 
1. Iterate back over the provResults, replacing flow/step if it doesn't exist.

Though - what's the requirement behind populating flow/step only if the flow/step still exists? If it's something for the UI, I think it would be better to add boolean properties of flowExists and stepExists. We shouldn't hide the flow/step name just because it no longer exists.